src/connection.rs

```diff
@@ -186,6 +186,88 @@ impl fmt::Debug for Connection {
     }
 }
 
+/// Whether an endpoint was the initiator of a connection
+#[cfg_attr(feature = "arbitrary", derive(Arbitrary))]
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub enum Side {
+    /// The initiator of a connection
+    Initiator = 0,
+    /// The acceptor of a connection
+    Acceptor = 1,
+}
+
+/// Whether a stream communicates data in both directions or only from the initiator
+#[cfg_attr(feature = "arbitrary", derive(Arbitrary))]
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub enum Dir {
+    /// Data flows in both directions
+    Bi = 0,
+    /// Data flows only from the stream's initiator
+    Uni = 1,
+}
+
+impl fmt::Display for Dir {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        use self::Dir::*;
+        f.pad(match *self {
+            Bi => "bidirectional",
+            Uni => "unidirectional",
+        })
+    }
+}
+
+/// Identifier for a stream within a particular connection
+#[cfg_attr(feature = "arbitrary", derive(Arbitrary))]
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct StreamId(#[doc(hidden)] pub u64);
+
+impl fmt::Display for StreamId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let initiator = match self.initiator() {
+            Side::Initiator => "initiator",
+            Side::Acceptor => "acceptor",
+        };
+        let dir = match self.dir() {
+            Dir::Uni => "uni",
+            Dir::Bi => "bi",
+        };
+        write!(
+            f,
+            "{} {}directional stream {}",
+            initiator,
+            dir,
+            self.index()
+        )
+    }
+}
+
+impl StreamId {
+    /// Create a new StreamId
+    pub fn new(initiator: Side, dir: Dir, index: u64) -> Self {
+        StreamId(index << 2 | (dir as u64) << 1 | initiator as u64)
```

Comment:
Why don't we simply use the enums instead of doing this type of encoding of these values?

Reply:
That was copied from quinn's design.
I think it's trying to encode the info into the u64

Reply:
We shouldn't need such encoding here I believe?